Cython/Compiler/ModuleNode.py

```diff
@@ -3400,7 +3414,13 @@ def generate_module_cleanup_func(self, env, code):
                 code.putln("while (%s > 0) {" % freecount_name)
                 code.putln("PyObject* o = (PyObject*)%s[--%s];" % (
                     freelist_name, freecount_name))
+                code.putln("#if !CYTHON_USE_TYPE_SLOTS && !CYTHON_COMPILING_IN_PYPY")
+                # Asking for PyType_GetSlot(..., Py_tp_free) seems to cause an error in pypy
+                code.putln("freefunc tp_free = (freefunc)PyType_GetSlot(Py_TYPE(o), Py_tp_free);")
+                code.putln("if (tp_free) tp_free(o);")
+                code.putln("#else")
                 code.putln("(*Py_TYPE(o)->tp_free)(o);")
+                code.putln("#endif")
```

Comment:
It would be better to spell both implementations, here and above, in the same order. I'd reverse the conditions here, not above. It seems more natural to spell out the specific cases followed by the generic fallback.